### PR TITLE
New ZestCLI (previously: "don't delete corpus on start")

### PR DESCRIPTION
**Pull request by yevgenypats**
So isn't that should be the default behaviour of not deleting the corpus on start?

I mean this way it will resume from the same place i.e will take random input from the corpus and mutate and continue?

maybe just throw a warning that the corpus directory is not empty and the fuzzing will continue from this corpus directory?

### CODE REVIEW

**Comment by yevgenypats** [comment]
Got it. I agree. I updated the commit with the following

* Add PicoCLI for ZestDriver as there are a lot of options and it is hard to remember and hard to use via -D java properties.
* The current best and user-friendly solution to the loading problem  I found is package the ZestDriver with assembly and release that jar independetly as well as package the target library with the fuzz tests. with this solution it works pretty nicely with:
`java -jar zest-driver.jar --help`
```
Usage: ZestDriver [-ehlV] [-i=<inputDirectory>] [-o=<outputDirectory>] PACKAGE
                  TEST_CLASS TEST_METHOD
      PACKAGE           package containing the fuzzer and all dependencies
      TEST_CLASS        full class name where the fuzz function is located
      TEST_METHOD       fuzz function name
  -e, --exit-on-crash   Exit fuzzer on first crash (default: false)
  -h, --help            Show this help message and exit.
  -i, --input=<inputDirectory>
                        Input directory containing seed test cases (default:
                          none)
  -l, --libfuzzer-compat-output
                        Use libFuzzer compat output instead of AFL like stats
                          screen (default: false)
  -o, --output=<outputDirectory>
                        Output Directory containing results (default:
                          fuzz_results)
  -V, --version         Print version information and exit.
```
and for fuzzing

`java -jar zest-driver.jar example-java-1.0-SNAPSHOT-fat-tests.jar dev.fuzzit.examplejava.ParseComplexTest fuzz`

this is kind of like afl usage and eliminates the need for the scripts + downloading the jqf repo.

Once that is reviewed adding the `-i-` mode should be easy

**Comment by rohanpadhye** [comment]
Hey this looks like a good idea but I see that all these changes to `ZestDriver.java` break backward compatibility in various ways. I have tons of scripts that rely on `bin/jqf-zest`, `bin/jqf-afl`, etc. working in their originally intended way. The use of InstrumentingClassLoader also means that JDK classes cannot be instrumented, which restricts usage to fuzzing application classes only.

Can you leave the original ZestDriver.java as it is and put this modified PicoCLI-based driver in a new class, e.g. `ZestPackagedDriver`? The name makes it clear that it has a different interface than all the other `XDrivers` (corresponding to `XGuidance`). You can use `ZestPackagedDriver` as the main class for the fat JAR built from the assembly plugin.

I'll add some review comments on specific lines of code.

**Comment by yevgenypats** [comment]
Cool, yes I'll do that as a different driver. Thank you for the quick reply!

File: fuzz/src/main/assembly/assembly.xml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+        <dependencySet>
+            <outputDirectory>/</outputDirectory>
+            <useProjectArtifact>true</useProjectArtifact>
+            <unpack>true</unpack>
+            <scope>test</scope>
+        </dependencySet>
+    </dependencySets>
+    <fileSets>
+        <fileSet>
+            <directory>${project.build.directory}/test-classes</directory>
```
[the remark this answers is not in the document]
Comment (yevgenypats) [reply fuzz/src/main/assembly/assembly.xml]: good catch. done.

File: fuzz/src/main/assembly/assembly.xml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    <formats>
+        <format>jar</format>
+    </formats>
+    <includeBaseDirectory>false</includeBaseDirectory>
+    <dependencySets>
+        <dependencySet>
+            <outputDirectory>/</outputDirectory>
+            <useProjectArtifact>true</useProjectArtifact>
+            <unpack>true</unpack>
+            <scope>test</scope>
```
[the remark this answers is not in the document]
Comment (yevgenypats) [reply fuzz/src/main/assembly/assembly.xml]: good catch done.

File: fuzz/src/main/assembly/assembly.xml [hunk 3]
```diff
@@ -0,0 +1,24 @@
+<assembly
+        xmlns="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/1.1.3"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/1.1.3 http://maven.apache.org/xsd/assembly-1.1.3.xsd">
+    <id>fat-tests</id>
```
[the remark this answers is not in the document]
Comment (yevgenypats) [reply fuzz/src/main/assembly/assembly.xml]: changed.

**Comment by yevgenypats** [comment]
@rohanpadhye I fixed the review per your comments and added an independent ZestCLI class that is based on PicoCLI and uses the class loader. let me know if more changes are necessary. if not let's merge this PR and I'll open another one from master that will implement the resume afl-like functionality.

File: fuzz/src/main/java/edu/berkeley/cs/jqf/fuzz/ei/ZestCLI.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+    @Option(names = { "-i", "--input" },
+            description = "Input directory containing seed test cases (default: none)")
+    private File inputDirectory;
+
+    @Option(names = { "-o", "--output" },
+            description = "Output Directory containing results (default: fuzz_results)")
+    private File outputDirectory = new File("fuzz-results");
+
+    @Parameters(index="0", paramLabel = "PACKAGE", description = "package containing the fuzzer and all dependencies")
```
Comment (rohanpadhye) [inline fuzz/src/main/java/edu/berkeley/cs/jqf/fuzz/ei/ZestCLI.java]: Should this be "package containing the fuzz *target* and all dependencies"?

**Pull request merged**